drivers/logserial.py

```diff
@@ -21,50 +21,38 @@
 
 
 class LogSerial(Thread):
-    def __init__(self, config, bus, com=None):
+    def __init__(self, config, bus):
         Thread.__init__(self)
         self.setDaemon(True)
         self.should_run = Event()
         self.should_run.set()
 
-        if com is None:
-            if 'port' in config:
-                self.com = serial.Serial(config['port'], config['speed'])
-                self.com.timeout = 0.01  # expects updates < 100Hz
-            else:
-                self.com = None
+        self.output_thread = Thread(target=self.run_output)
+
+        if 'port' in config:
+            self.com = serial.Serial(config['port'], config['speed'])
+            self.com.timeout = 0.01  # expects updates < 100Hz
         else:
-            self.com = com
+            self.com = None
         self.bus = bus
 
         self.buf = b''
 
+    def start(self):
+        Thread.start(self)
+        self.output_thread.start()
+
+    def join(self, timeout=None):
+        Thread.join(self, timeout=timeout)
+        self.output_thread.join(timeout=timeout)
+
     def run(self):
         while self.should_run.isSet():
             data = self.com.read(1024)
             if len(data) > 0:
                 self.bus.publish('raw', data)
```

Comment:
Could we make `publish` to throw `BusShutdownException` so that we do not have to special case this? Or is there any other reason to keep it this way?

Reply:
well, it still should terminate even there are no data incoming, i.e. nothing to publish??

Reply:
Hmm, you are right. But I still somehow think that shutdown info should come from the bus. This way we have 2 different ways to cause shutdown. That does not seem right.

Reply:
> There should be one-- and preferably only one --obvious way to do it.

[The Zen of Python](https://www.python.org/dev/peps/pep-0020/)

Reply:
We could add something like
```
while self.bus.alive():
     <main loop>
```
The `bus` currently empties the whole input queue before it is terminated ...

Reply:
👍